pkg/csi/service/wcpguest/controller_helper.go

```diff
@@ -213,9 +215,82 @@ func getPersistentVolumeClaimSpecWithStorageClass(pvcName string, namespace stri
 			StorageClassName: &storageClassName,
 		},
 	}
+	snapshotApiGroup := common.VolumeSnapshotApiGroup
+	volumeSnapshotKind := common.VolumeSnapshotKind
+	if volumeSnapshotName != "" {
+		localObjectReference := &v1.TypedLocalObjectReference{
+			APIGroup: &snapshotApiGroup,
+			Kind:     volumeSnapshotKind,
+			Name:     volumeSnapshotName,
+		}
+		claim.Spec.DataSource = localObjectReference
+	}
 	return claim
 }
 
+func getVolumeSnapshotWithVolumeSnapshotClass(volumeSnapshotName string, namespace string,
+	volumeSnapshotClassName string, pvcName string) *snap.VolumeSnapshot {
+	volumeSnapshot := &snap.VolumeSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      volumeSnapshotName,
+			Namespace: namespace,
+		},
+		Spec: snap.VolumeSnapshotSpec{
+			Source: snap.VolumeSnapshotSource{
+				PersistentVolumeClaimName: &pvcName,
+			},
+			VolumeSnapshotClassName: &volumeSnapshotClassName,
+		},
+		Status: nil,
+	}
+	return volumeSnapshot
+}
+
+// isVolumeSnapshotInSupervisorClusterReady return true if the VolumeSnapshot is ReadyToUse
+func isVolumeSnapshotInSupervisorClusterReady(ctx context.Context, client snapshotterClientSet.Interface,
```

Comment:
Is this function specific for Supervisor Cluster?  Do we already have a similar one for Vanilla?

Reply:
It's not specific to supervisor actually, since the parameter can take in vanilla snapshotter `client` and do the same check for vanilla, but, there is no use to make it generic as we don't check for VolumeSnapshot status in vanilla.

Reply:
It's good to move this to a generic place.

Reply:
Fixed.